### DISCUSSION TITLE
Test and response assertions

### DISCUSSION THREAD

**Discussion by jacek-kurlit** [LucasPickering/slumber#272] [Ideas]
Hi,

First of all I want to tell you that this is great project!

I have question about asserting responses
Post man has this option to check if response passes some kind of assertions like status == 200 etc
Will there be such feature in slumber too?

**Comment by LucasPickering** [comment]
I'm confused about the application of this feature. I've never used Postman but the quick googling I just did only turned up results about assertions in test scripts. Is that what you're talking about, or are you referring to assertions within the UI? If the latter, what purpose do the assertions serve?

**Reply by jacek-kurlit** [reply]
Assertions allows to validate if output of result returned by server is what you have expected.
So in postman you could write assert_eq!(response.status, 200) or something like that.
This allows to create test suites for e2e test and run them all.

I'm not sure if such feature is needed at all, probably it may be overkill since there are tools that allows to e2e tests like [hurl](https://hurl.dev/)
I'm just curious of general direction this project will take (like becoming terminal postman etc)

**Reply by LucasPickering** [reply]
While I have no plans to build any features specific for e2e testing, you could easily use the existing CLI functionality for some limited testing. For example, to run a request and fail if the status code is >= 400:

```
slumber request createUser --exit-status
```

There are other flags to control what is printed to the console, which you could integrate into some sort of script to compare. This is clunky though so I wouldn't recommend it for any meaningful tests.

I would potentially be open to distributing the core Slumber functionality as its own crate with a library interface, which would allow you to write Rust code to build test suites, but that's a large commitment to maintaining a stable interface so I don't plan to do it soon.
